File: src/models/population/Population.ts
import TSP from "../TSP";
import Individual from "../Individual";

export default abstract class Population {
    public individuals: Individual[];

    constructor(public size: number, public tsp: TSP) {
        if (!tsp) {
            this.individuals = [];
            return;
        }

        if (size > tsp.dimension) this.size = tsp.dimension;

        this.individuals = new Array<Individual>(this.size)
            .fill(undefined!)
            .map(() => new Individual(this.tsp.towns));

        this.initialize();
    }

    protected abstract initialize(): void;

    public calculateBestDistance(): number {
        const distances = this.individuals.map(individual =>
            individual.calculateTotalDistance()
        );

        return Math.min(...distances);
    }

    public calculateAverageDistance(): number {
        const distances = this.individuals.map(individual =>
            individual.calculateTotalDistance()
        );

        return distances.reduce((a, b) => a + b, 0) / distances.length;
    }

    public calculateWorstDistance(): number {
        const distances = this.individuals.map(individual =>
            individual.calculateTotalDistance()
        );

        return Math.max(...distances);
    }
}

File: zad1/Models/Population.cs
using System.Collections.Generic;
using System.Linq;

namespace zad1.Models
{
    public abstract class Population
    {
        public List<Individual> Individuals { get; set; }
        protected int Size { get; }
        private TSP Tsp { get; }

        protected Population(int size, TSP tsp)
        {
            Size = size;
            Tsp = tsp;

            Individuals = new List<Individual>();

            for (var i = 0; i < size; i++)
                Individuals.Add(new Individual(tsp.Towns));

            Initialize();
        }

        protected abstract void Initialize();

        public double CalculateBestDistance()
        {
            return Individuals.Select(individual => individual.CalculateTotalDistance()).Min();
        }

        public double CalculateAverageDistance()
        {
            return Individuals.Select(individual => individual.CalculateTotalDistance()).Average();
        }

        public double CalculateWorstDistance()
        {
            return Individuals.Select(individual => individual.CalculateTotalDistance()).Max();
        }
    }
}